app/controllers/admin/categories_controller.rb

```diff
@@ -0,0 +1,48 @@
+class Admin::CategoriesController < ApplicationController
+  before_action :authenticate_user!
+  before_action :set_category, except: [:index, :new, :create]
+
+  def index
+    @categories= ProductCategory.all
+  end
+
+  def new
+    @category = ProductCategory.new
+  end
+
+  def create
+    @category = ProductCategory.new(category_params)
+    @category.save!
+    redirect_to admin_categories_path
+  end
+
+  def show; end
+
+  def edit; end
+
+  def update
+    if @category.update(category_params)
+      flash[:notice] = 'Category updated!'
+      redirect_to admin_categories_path
+    else
+      flash[:error] = 'Failed to edit Category!'
```

Comment:
just render edit and show the object errors in form, google how to show the model object error messages and then write that code in a partial in a generic way and save it in shared folder and use it in your forms